compiler/p/codegen/ControlFlowEvaluator.cpp

```diff
@@ -3001,7 +3001,7 @@ static void lookupScheme3(TR::Node *node, bool unbalanced, TR::CodeGenerator *cg
          TR_PPCTableOfConstants::setTOCSlot(offset, address);
          if (offset<LOWER_IMMED||offset>UPPER_IMMED)
             {
-            generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addis, node, addrRegister, cg->getTOCBaseRegister(), cg->hiValue(offset));
+            generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addis, node, addrRegister, cg->getTOCBaseRegister(), (int16_t)cg->hiValue(offset));
```

Comment:
I think hiValue() returns a value that can be bigger than 16-bit:

// These macros are intended for use when HI_VALUE and LO_VALUE will be recombined after LO_VALUE is sign-extended
// (e.g. when LO_VALUE is used with an instruction that takes a signed 16-bit operand).
// In this case we have to adjust HI_VALUE now so that the original value will be obtained once the two are recombined.
#define HI_VALUE(x)               (((x)>>16) + (((x)>>15)&1))
#define LO_VALUE(x)               ((int16_t)(x))

Are we sure we want to truncate it here?

Reply:
`offset` is an `int32_t`. Calling `hiValue` will return something that can fit in 16 bits.

Reply:
As long as hiValue is signed it's ok I think. I would rather minimize the changes. @aviansie-ben what do you think?

Reply:
Actually, after thinking on this a bit longer I realized that truncating the return value of `HI_VALUE` is not a proper fix, even if the offset is guaranteed to be 32 bits or less. Take for instance an offset of `0x7fffffff`. By truncating the high value, we would end up with the following code:

```
addis rX, toc, -0x8000
ld rY, -1(rX)
```

This actually results in an offset of `0xffffffff7fffffff`, which is completely incorrect. I'm curious as to how we'd get an immediate this high in the first place but regardless of that, the return value _should not be truncated_, otherwise the resulting code might be incorrect without warning that something bad happened.

Reply:
That's a good catch. Truncating the value would just cover up the problem there. I'll remove it from the locations that cause a problem like this.

Reply:
that is a moot point anyway.  pTOC is never designed to be over 4GB. and, using the existing sequence (addis/ld), truncating or not doesn't matter except cosmetic assertion on that value.  the consuming by addis doesn't change.

Reply:
pTOC base is at the middle of the TOC, if you are interested in the possible offsets.

Reply:
I changed the code to specifically detect cases like `0x7FFFFFFF`. It now generates code like this:

```
addis rX, toc, 0x7FFF
addis rX, rX, 0x1
ld rY, 0xFFFF[rX]
```
Prior to any of my changes, it would have just triggered the assert. Without the assert, it would have generated incorrect code.

Reply:
@IBMJimmyk who adds the second instruction? I thought we simply issue two instructions that might be incorrect in some border cases, but then we have an ASSERT for such cases.

Reply:
The second `addis` was added by my changes. Although, it was removed again in most places in later changes.